File: C++/PBKDF2Helper.cpp
#include <iostream>
#include <stdexcept>
#include "PBKDF2Helper.h"
#include "algparam.h"
#include "pwdbased.h"
#include "sha.h"

PBKDF2Helper::PBKDF2Helper(std::string pwd, std::vector<byte> salt, int iterations = 5000)
	: m_iterations(iterations)
	, m_salt(salt)
{
	this->m_password = std::vector<byte>(pwd.begin(), pwd.end());
}

bool PBKDF2Helper::GetBytes(unsigned int count, std::vector<byte>& derivedBytes)
{
	std::cout << "PBKDF2 algorithm to get " << count << " bytes started" << std::endl;

	if (count > 0)
	{
		derivedBytes.clear();
		derivedBytes.resize(count);

		CryptoPP::PKCS5_PBKDF2_HMAC<CryptoPP::SHA512> pbkdf2;

		byte *derived = derivedBytes.data();
		const byte *pwd = this->m_password.data();
		const byte *salt = this->m_salt.data();

		try
		{
			pbkdf2.DeriveKey(derived, count, 0, pwd, this->m_password.size(), salt, this->m_salt.size(), this->m_iterations);
		}
		catch (const std::exception&)
		{
			std::cerr << "Could not derive bytes with PBKDF2" << std::endl;
			throw;
		}

		std::cout << "PBKDF2 algorithm finished" << std::endl;
		return true;
	}
	else
	{
		throw std::runtime_error("Parameter 'count' can't be zero");
	}
}

File: Java/PBKDF2Helper.java
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.charset.Charset;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.KeySpec;
import java.util.Arrays;

class PBKDF2Helper {
    private final int iterations;
    private final byte[] salt;
    private final byte[] password;

    public PBKDF2Helper(final String pwd, final byte[] salt, final int iterations) {
        this.iterations = iterations;
        this.salt = salt;
        this.password = pwd.getBytes();
    }

    public byte[] getBytes(int count) {
        System.out.println("PBKDF2 algorithm to get " + count + " bytes started");

        if (count < 0) {
            throw new RuntimeException("Parameter 'count' can't be zero or smaller");
        }

        byte[] result;
        try {
            Charset charset = Charset.forName("UTF-8");
            CharBuffer charBuffer = charset.decode(ByteBuffer.wrap(this.password));
            char[] input = Arrays.copyOf(charBuffer.array(), this.password.length);

            KeySpec spec = new PBEKeySpec(input, this.salt, this.iterations, count * 8);
            SecretKeyFactory keyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512");

            result = keyFactory.generateSecret(spec).getEncoded();
        } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
            throw new RuntimeException("Could not derive bytes with PBKDF2", e);
        }

        System.out.println("PBKDF2 algorithm finished");

        return result;
    }
}
